### PR TITLE
Support PyTorch models

### PR DESCRIPTION
**Pull request by josebummer**
- Added support for PyTorch models.
- Added notebook using a PyTorch model.
- Added test for 100% coverage

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -11,6 +11,6 @@
       long_description_content_type="text/markdown",
       url="https://github.com/sherpaai/Sherpa.ai-Federated-Learning-Framework",
       packages=find_packages(),
-      install_requires=['numpy', 'emnist', 'scikit-learn>=0.23', 'pytest', 'tensorflow>=2.2.0', 'scipy', 'six', 'pathlib2',
-                        'pandas', 'multipledispatch'],
+      install_requires=['numpy', 'emnist', 'scikit-learn>=0.23', 'pytest', 'pytest-cov', 'tensorflow>=2.2.0', 'scipy', 'six', 'pathlib2', 'torch>=1.0',
+                        'pandas', 'multipledispatch', 'Keras'],
```
Comment (gegose) [inline setup.py:15]: Why we need explicit Keras dependency? We require tensorflow>=2.2.0 and keras is included in tensorflow since 2.0, right?

Comment (josebummer) [reply setup.py:15]: Yes, you are right, my fault. I'm going to change it.

File: shfl/model/deep_learning_model_pt.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    # Arguments:
+        model: Compiled model, ready to train
+        criterion: Loss function to apply
+        optimizer: Optimizer to apply
+        batch_size: batch_size to apply
+        epochs: Number of epochs
+        metrics: Metrics for apply. Dictionary {name: function to apply, ...}. Default shows loss and accuracy
+        device: Device where it will run. Default cpu
+    """
+    def __init__(self, model, criterion, optimizer, batch_size=32, epochs=1, metrics=None, device="cpu"):
```
Comment (gegose) [inline shfl/model/deep_learning_model_pt.py:21]: This interface differs a little from DeepLearningModel class. I know the type of objects that receives each one are different but the meaning is the same. I think this format could be better (for example __deepcopy__ method could be cleaner) but i'm not completely sure if modify DeepLearningModel to comply with this kind of signature is trivial. What do you think? Do we create a task to adapt DeepLearningModel to the same format? This would affect lots of notebooks, of course.

Comment (josebummer) [reply shfl/model/deep_learning_model_pt.py:21]: We could modify `DeepLearningModel` class to pass it the loss function, optimizer and metrics by parameters and then make the `model.compile()` inside the class. The `device` parameter is not adaptable, since to run on GPU, TF and PyTorch work differently and that parameter is specifically for that.

With these changes, we may have to modify the `deepcopy` method to apply the `model.compile()` when copying a model. First of all, we should copy all the attributes and then perform the `model.compile()`.

So we can create a task to adapt `DeepLearningModel` class to the same format. I think that the task would be composed of:

1. Modify `__init__()` and `deepcopy` methods for DeepLearningModel class (I'm not sure if more methods are necessary).
2. Modify all notebooks where `DeepLearningModel`class appears.
3. Modify tests for obtain 100% coverage.

Regards.

Comment (gegose) [reply shfl/model/deep_learning_model_pt.py:21]: Exactly, the idea was exactly like you have described. I'm going to create this issue.

**Review by gegose — APPROVED** [review]
